From c397ab2a3bc9b9568b6e98dc7b90f5ffccc3c814 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Mon, 31 Mar 2014 17:26:16 -0400
Subject: [net] ipv6: reallocate addrconf router for ipv6 address when lo device up

Message-id: <08e58ccdc84f0bc7557bdd16468d5dfbf57915e0.1396286769.git.jbenc@redhat.com>
Patchwork-id: 78723
O-Subject: [RHEL7.1 net PATCH] ipv6: reallocate addrconf router for ipv6 address when lo device up
Bugzilla: 1077053
RH-Acked-by: Thomas Graf <tgraf@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1077053
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7277764
Tested: Using the reproducer from bugzilla

commit 33d99113b1102c2d2f8603b9ba72d89d915c13f5
Author: Gao feng <gaofeng@cn.fujitsu.com>
Date:   Fri Jan 24 16:29:11 2014 +0800

    ipv6: reallocate addrconf router for ipv6 address when lo device up

    commit 25fb6ca4ed9cad72f14f61629b68dc03c0d9713f
    "net IPv6 : Fix broken IPv6 routing table after loopback down-up"
    allocates addrconf router for ipv6 address when lo device up.
    but commit a881ae1f625c599b460cc8f8a7fcb1c438f699ad
    "ipv6:don't call addrconf_dst_alloc again when enable lo" breaks
    this behavior.

    Since the addrconf router is moved to the garbage list when
    lo device down, we should release this router and rellocate
    a new one for ipv6 address when lo device up.

    This patch solves bug 67951 on bugzilla
    https://bugzilla.kernel.org/show_bug.cgi?id=67951

    change from v1:
    use ip6_rt_put to repleace ip6_del_rt, thanks Hannes!
    change code style, suggested by Sergei.

    CC: Sabrina Dubroca <sd@queasysnail.net>
    CC: Hannes Frederic Sowa <hannes@stressinduktion.org>
    Reported-by: Weilong Chen <chenweilong@huawei.com>
    Signed-off-by: Weilong Chen <chenweilong@huawei.com>
    Signed-off-by: Gao feng <gaofeng@cn.fujitsu.com>
    Acked-by: Hannes Frederic Sowa <hannes@stressinduktion.org>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv6/addrconf.c b/net/ipv6/addrconf.c
index b0f5b41..cf25455 100644
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -2635,8 +2635,18 @@ static void init_loopback(struct net_device *dev)
    if (sp_ifa->flags & (IFA_F_DADFAILED | IFA_F_TENTATIVE))
     continue;
 
-   if (sp_ifa->rt)
-    continue;
+   if (sp_ifa->rt) {
+    /* This dst has been added to garbage list when
+     * lo device down, release this obsolete dst and
+     * reallocate a new router for ifa.
+     */
+    if (sp_ifa->rt->dst.obsolete > 0) {
+     ip6_rt_put(sp_ifa->rt);
+     sp_ifa->rt = NULL;
+    } else {
+     continue;
+    }
+   }
 
    sp_rt = addrconf_dst_alloc(idev, &sp_ifa->addr, 0);
 
-- 
1.7.1